### DISCUSSION TITLE
v0.10.2-beta1 2023-12-19

### DISCUSSION THREAD

**Discussion by nimroddolev** [nimroddolev/chime_tts#59] [Announcements]
## ✨ Features
* Added `FFmpeg Arguments` parameter to the `chime_tts.say` and `chime_tts.say_url` services, allowing FFmpeg conversions of the generated audio.
* Added support for FFmpeg audio conversion of generated audio.
* Added audio `duration` value to `chime_tts.say` service response value object.

## 🧰 Improvements
* Added fix for double back slashes in user configured folder paths.

## 🐞 Bugfixes
* Fixed default service parameter values.

## What's Changed
* Bump actions/setup-python from 4.7.1 to 5.0.0 by @dependabot in https://github.com/nimroddolev/chime_tts/pull/55
* Bump ruff from 0.1.6 to 0.1.7 by @dependabot in https://github.com/nimroddolev/chime_tts/pull/54
* Bump ruff from 0.1.7 to 0.1.8 by @dependabot in https://github.com/nimroddolev/chime_tts/pull/58

**Full Changelog**: https://github.com/nimroddolev/chime_tts/compare/v0.10.1...v0.10.2-beta1

<hr /><em>This discussion was created from the release <a href='https://github.com/nimroddolev/chime_tts/releases/tag/v0.10.2-beta1'>v0.10.2-beta1 2023-12-19</a>.</em>

**Comment by nimroddolev** [comment]
This beta release introduces support for running FFmpeg conversion jobs on the audio generate by the `chime_tts.say` and `chime_tts.say_url` services, as well as some other improvements.

![Screenshot 2023-12-19 at 15 21 34](https://github.com/nimroddolev/chime_tts/assets/1849295/2fabaf53-439e-4192-9882-785e5579d529)

The main use case for the new FFmpeg support is for playback on 3rd party media_player entities such as Alexa and Google Home speakers.

I don't personally own any of these speakers so I can't test on actual hardware, but for those of you who do I would appreciate any feedback you might have on the feature, and in general as well.

To install the beta:
1. Open the <a href="https://my.home-assistant.io/redirect/hacs_repository/?owner=nimroddolev&repository=chime_tts&category=integration" title="Chime TTS HACS repository" target="_blank">Chime TTS repository in HACS</a>
1. Click on the `⋮` button in the top right
1. Select the `↻ Redownload` option
1. Enable the `Show beta versions` option
1. Select version `v0.10.2-beta1` from the updated version drop-down list

Thanks!

**Reply by dbullendsl** [reply]
the final output is getting dropped into the wrong location, instead of /www/chime_tts it's just going into /www

also, I think the FFMPEG check box should just fill with default settings, rather than having a user try to figure them out.

**Reply by nimroddolev** [reply]
> the final output is getting dropped into the wrong location, instead of /www/chime_tts it's just going into /www

Did you check the value of the public mp3 folder in the integration's configuration? Eg:

![Screenshot 2023-12-19 at 17 56 23](https://github.com/nimroddolev/chime_tts/assets/1849295/9e40e075-bb10-4340-a09e-1110bf5ade53)

> also, I think the FFMPEG check box should just fill with default settings, rather than having a user try to figure them out.

I can't pre-fill the field with a value (due to a limitation of services in Home Assistant). What I can do is add a drop down list for standard conversions (such as Alexa, Google Home, etc.) and also include the free-form field (as implemented in the beta) to provide the best of both worlds.

**Reply by dbullendsl** [reply]
ok, it's working with Alea - looks like you changed what that field should be, I didn't have the trailing backslash but it worked before the installation of the beta. After the beta install my integration broke, put in the backslash and all is good.

btw - the params I used in the ffmpeg field are: 

```
 -y -ac 2 -codec:a libmp3lame -b:a 48k -ar 24000 -write_xing 0
```

**Reply by nimroddolev** [reply]
> looks like you changed what that field should be, I didn't have the trailing backslash but it worked before the installation of the beta

I am not sure what changed...do you remember what the value was before you changed it? Also - do you happen to remember which version of Chime TTS you were running before the beta?

**Reply by dbullendsl** [reply]
my old setting was /config/www/chime_tts  not a big difference, I must have been running the most recent release version as I had only recently discovered your integration.

**Reply by nimroddolev** [reply]
Thank you. I fixed the issue for the next release

**Reply by dbullendsl** [reply]
Ok, no problem - I'm trying to figure out how to pass options to amazon_polly to get the voice changed - no luck so far. Anything I've tried results in having to either reload the integration or restarting home assistant as the tts service stops.

**Reply by nimroddolev** [reply]
I don't think it's possible.

As I understand it, the voice for Amazon Polly is part of Amazon Polly's configuration in your configuration.yaml, and not a parameter you can include in the Options when sending off the request for the TTS audio.
I thought of creating multiple Amazon Polly tts platforms, each with a different voice configuration, but I don't think you can have more than 1 instance of Amazon Polly in Home Assistant.

If you find something to the contrary let me know, i'm also interested.

**Reply by dbullendsl** [reply]
I'm sure you're right, although Amazon AWS alludes to the idea it can be done with their Alexa Tool Kit. But that's way beyond the scope of what you're trying to do. For now I'm going to stick to using the notify.alexa_media service call.

**Comment by KEV-github** [comment]
I have been trying to get this to work with Alexa.
See my config attached
Alexa is replying with to send tts set up my public url in the intergration configuration. Any ideas what ive missed.
<img width="1361" alt="Screenshot 2023-12-20 at 17 35 18" src="https://github.com/nimroddolev/chime_tts/assets/57045329/1b936f17-9b90-4300-98d6-b8b4cbd2c00f">

**Reply by nimroddolev** [reply]
Hi @KEV-github. 

Alexa speakers will only play MP3s from publicly accessible URLs (so not local files on). So you need to make sure your Home Assistant instance is accessible remotely. Here's [a guide with a few options to do this](https://www.wundertech.net/how-to-access-home-assistant-remotely/) in case you haven't already.

Once that's configured, you simply need to enter your external address to your HA instance into the Amazon Media Player integration's configuration. Go to `Settings` --> `Devices & Services` --> `Alexa Media Player` --> `Configure`  and then enter the external URL to your Home Assistant instance.

I don't own any Alexa speakers, so perhaps @dbullendsl you could verify this is correct?

**Reply by dbullendsl** [reply]
the chime path looks wrong, look at your settings these work for me

![Capture](https://github.com/nimroddolev/chime_tts/assets/98482191/e30abf73-1013-46f8-84f1-238e5af55945)

here's a script that works:

```
alias: call chime tts
sequence:
  - service: chime_tts.say_url
    data:
      chime_path: bells
      end_chime_path: null
      delay: 0
      final_delay: 0
      tts_playback_speed: 100
      tts_platform: google_translate
      cache: true
      ffmpeg_args: " -y -ac 2 -codec:a libmp3lame -b:a 48k -ar 24000 -write_xing 0"
      message: "{{message}}"
    response_variable: chime_tts
  - service: notify.alexa_media
    data:
      message: <audio src= "{{chime_tts.url}}"/>
      target:
        - media_player.office_echo_dot
      data:
        type: tts
mode: single
fields:
  message:
    selector:
      text: null
    name: message
```
you can test this by going to developer tools, under services and enter service call to script:call_chime_tts, check the Message box and enter your message.

**Reply by nimroddolev** [reply]
I think you might have an issue with your configuration values. I was successful:

---

My configuration:

<img width="510" alt="Screenshot 2023-12-21 at 0 48 33" src="https://github.com/nimroddolev/chime_tts/assets/1849295/b92c6d10-7989-4f97-8dfb-0a0375bee74b">

---

The result of the `chime_tts.say_url` service call:

<img width="679" alt="Screenshot 2023-12-21 at 0 50 44" src="https://github.com/nimroddolev/chime_tts/assets/1849295/423bec74-7f22-48fd-a75e-314e46462df7">

**Reply by dbullendsl** [reply]
one other thing I was wondering, the address in your URL looks like a local URL not available publically, as far as I recall Amazon requires a public URL. That means getting a SSL certificate also - look into the DuckDNS addin in the add in store - that can get you a public URL in the duckdns.org domain and SSL certificates so then Amazon access your server with proper credentials. It's all free - so it's just the matter of getting it set up.

**Reply by nimroddolev** [reply]
> the address in your URL looks like a local URL not available publically

No, I obscured my domain but I can assure you the address is publicly available.

> as far as I recall Amazon requires a public URL. That means getting a SSL certificate also

Yes, that’s exactly what I wrote in [my comment](https://github.com/nimroddolev/chime_tts/discussions/59#discussioncomment-7911562)

**Comment by KEV-github** [comment]
I have been reading through your reply's and set up my external Nabu Casa url in the amazon intergration. Am I Correct in saying this should now let me play mp3 and tts on my alexa's.?
I have also changed my Chime config as shown above.

**Reply by nimroddolev** [reply]
Yes, is it working for you?

**Reply by KEV-github** [reply]
this what I get with a basic call service
![Image 21-12-2023 at 19 23](https://github.com/nimroddolev/chime_tts/assets/57045329/c82add88-3be1-4a5b-aeda-2d7ca19047c0)

**Reply by KEV-github** [reply]
The mp3 file it creates is showing in my /local/Chime_tts file

**Reply by KEV-github** [reply]
Just need to work out the yaml config to play on my Alexa . Notify or media_player play

**Reply by nimroddolev** [reply]
The YAML looks correct, but Chime TTS only sees your local address, not your external one. Perhaps you need to set your external address in the settings?

Go to `Settings` --> `System`--> `Network` and scroll down to the `Home Assistant URL` section.
You should set the `Internet` address to your Nabu Casa URL.

**Comment by KEV-github** [comment]
Yes its set to This
![Image 21-12-2023 at 19 39](https://github.com/nimroddolev/chime_tts/assets/57045329/2a278a1a-0d26-4d03-894c-479bb60ce0e1)

**Comment by KEV-github** [comment]
if I use this call service it plays the chime and the message
![Image 21-12-2023 at 20 03](https://github.com/nimroddolev/chime_tts/assets/57045329/685c4ed7-cbf7-40b8-9526-dc2bff7cd987)

**Comment by nimroddolev** [comment]
Good to hear, but the URL from `chime_tts.say_url` is incorrect in your case. I can't reproduce the issue but i'm investigating.

**Reply by nimroddolev** [reply]
@KEV-github I have release a new beta version, v0.10.2-beta2 which I hope fixes your external URL issue (among other improvements).

Please upgrade and [report in this discussion](https://github.com/nimroddolev/chime_tts/discussions/62) if it solved your issue.

To install the beta:
1. Open the <a href="https://my.home-assistant.io/redirect/hacs_repository/?owner=nimroddolev&repository=chime_tts&category=integration" title="Chime TTS HACS repository" target="_blank">Chime TTS repository in HACS</a>
1. Click on the `⋮` button in the top right
1. Select the `↻ Redownload` option
1. Enable the `Show beta versions` option
1. Select version `v0.10.2-beta2` from the updated version drop-down list

**Comment by KEV-github** [comment]
@dbullendsl

Can you check this script what I copied from above. 'Im getting error messages .Thanks
![Image 22-12-2023 at 10 43](https://github.com/nimroddolev/chime_tts/assets/57045329/1a0b3181-1e22-46ae-bbbe-7879559ab41f)

**Reply by nimroddolev** [reply]
The parameter `ffmpeg_args` was renamed in beta2 to `audio_conversion`. It should work once you change the name.
Can you please try to use the in-built "Alexa" option? Eg:
```
alias: call chime tts
sequence:
  - service: chime_tts.say_url
    data:
      chime_path: bells
      tts_platform: google_translate
      cache: true
      audio_conversion: "Alexa”
      message: "{{message}}"
    response_variable: chime_tts
  - service: notify.alexa_media
    data:
      message: <audio src= "{{chime_tts.url}}"/>
      target:
        - media_player.kevin_s_echo_zibgee
      data:
        type: tts
mode: single
fields:
  message:
    selector:
      text: null
    name: message
```

**Reply by KEV-github** [reply]
![Image 22-12-2023 at 12 24](https://github.com/nimroddolev/chime_tts/assets/57045329/f17c7d3c-1c11-4a93-83d4-47111fee9347)
Still getting error message.

**Reply by nimroddolev** [reply]
Is the error for the service call to `chime_tts.say_url` or to `notify.alexa_media` ?
You can remove the second service call to see which one is causing the error.
@dbullendsl are you able to please see if you have the same issue?

**Comment by dbullendsl** [comment]
I'm not getting the same error, but I do get an error that the response_variable [chime_tts] is undefined from within the section trying to play on the alexa. My trace on the script doesn't show any value for '{{chime_tts.url}}'

When I simply call the service chime_tts.say_url I do get a response variable as a return that is correct.

**Reply by nimroddolev** [reply]
I think there is an accidental space character between the ‘=‘ and `"{{chime_tts.url}}" `:

 >```<audio src= "{{chime_tts.url}}"/> ```

Could that be it?

**Comment by KEV-github** [comment]
I dot know what the issue was.  Maybe the copy and paste didn't like it. So I created it using the visual editor and it worked.
It looks exactly the same though🤷🏻

![Image 22-12-2023 at 13 58](https://github.com/nimroddolev/chime_tts/assets/57045329/e3eccef1-8b79-4164-812b-6bc4cc9d040a)

**Reply by nimroddolev** [reply]
Got it. Glad it’s finally working 👍

**Reply by KEV-github** [reply]
Great work and thanks. Is there any way to make the chime longer and the speech louder. 
Thanks. Kevin

**Reply by nimroddolev** [reply]
You can add your own chimes (see the wiki) and you change use the `volume` parameter to adjust the volume.
